lib/pool/wait_queue.js

```diff
@@ -0,0 +1,136 @@
+'use strict';
+
+const errors = require('./errors');
+const PoolClosedError = errors.PoolClosedError;
+const WaitQueueFullError = errors.WaitQueueFullError;
+const WaitQueueTimeoutError = errors.WaitQueueTimeoutError;
+
+class WaitQueueMember {
+  constructor(callback) {
+    this.callback = callback;
+    this.finished = false;
+    this.timeout = null;
+  }
+
+  _finish(err, ret) {
+    if (!this.finished) {
+      this.finished = true;
+      process.nextTick(() => this.callback.call(null, err, ret));
+    }
+
+    if (this.timeout) {
+      clearTimeout(this.timeout);
+    }
+  }
+
+  success(connection) {
+    this._finish(null, connection);
+  }
+
+  failure(err) {
+    this._finish(err);
+  }
+
+  setTimeout(cb, ms) {
+    this.timeout = setTimeout(cb, ms);
+  }
+}
+
+class WaitQueue {
+  constructor(options) {
+    this._destroyed = false;
+
+    this.timeoutMS = options.waitQueueTimeoutMS || null;
```

Comment:
If `options.waitQueueTimeoutMS` isn't set, will `this.timeoutMS = options.waitQueueTimeoutMS` result in `this.timeoutMS = null`?

Reply:
yes. Though we could also set it to 0?

Reply:
I'm in favor of setting it to `0` so it matches the default for `options.waitQueueTimeoutMS`.

Reply:
good point, I'll update it.